felix/fv/policy_grouping_test.go

```diff
@@ -0,0 +1,196 @@
+						Selector: w[i].NameSelector(),
+					},
+				},
+			}
+			pol.Spec.Egress = []v3.Rule{
+				{
+					Action: actions[i],
+					Destination: v3.EntityRule{
+						Selector: w[i].NameSelector(),
+					},
+				},
+			}
+			_, _ = client.GlobalNetworkPolicies().Delete(context.TODO(), pol.Name, options.DeleteOptions{})
+			pol, err := client.GlobalNetworkPolicies().Create(context.TODO(), pol, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}
+
+	It("should implement policy semantics correctly independent of grouping", func() {
+		// Two selectors that are functionally equivalent here (they match all
+		// endpoints) but they will trigger different policy groupings.
+		const selA = "all()"
+		const selB = "has(name)"
+
+		for _, testCase := range []struct {
+			Selectors        []string
+			ExpectGroupChain bool
+		}{
+			// One big group.
+			{[]string{selA, selA, selA}, true},
+			// Group group with two policies, one inlined.
+			{[]string{selA, selA, selB}, true},
+			// Group inlined group followed by group with two.
+			{[]string{selA, selB, selB}, true},
+			// No groups due to alternating selector.
+			{[]string{selA, selB, selA}, false},
+		} {
+			// For each pattern of selectors (which should give a different
+			// policy grouping), move the deny rule between the three policies
+			// so that we know that all three policies are active and working.
+			for epToDeny := range w {
+				By(fmt.Sprintf("Grouping policies on selectors %v and denying traffic to/from WEP %d", testCase.Selectors, epToDeny))
+				actions := []v3.Action{v3.Allow, v3.Allow, v3.Allow}
+				actions[epToDeny] = v3.Deny
+				createPolicies(testCase.Selectors, actions)
+				for fromEPIdx := range w {
+					toEPIdx := (fromEPIdx + 1) % len(w)
+					expectation := connectivity.Some
+					if fromEPIdx == epToDeny || toEPIdx == epToDeny {
+						expectation = connectivity.None
+					}
+					cc.Expect(expectation, w[fromEPIdx], w[toEPIdx])
+				}
+				cc.CheckConnectivity()
+				cc.ResetExpectations()
+
+				iptablesChains := tc.Felixes[0].IPTablesChains("filter")
+				numInboundGroups := 0
+				numOutboundGroups := 0
+				for chainName := range iptablesChains {
+					if strings.HasPrefix(chainName, rules.PolicyGroupInboundPrefix) {
+						numInboundGroups++
+					} else if strings.HasPrefix(chainName, rules.PolicyGroupOutboundPrefix) {
+						numOutboundGroups++
+					}
+				}
+				if testCase.ExpectGroupChain {
+					Expect(numOutboundGroups).To(Equal(1),
+						"expected exactly one outbound group chain")
+					Expect(numInboundGroups).To(Equal(1),
+						"expected exactly one inbound group chain")
+				} else {
+					Expect(numOutboundGroups).To(Equal(0),
+						"expected no outbound group chains (failed cleanup of previous scenario?)")
+					Expect(numInboundGroups).To(Equal(0),
+						"expected no inbound group chains (failed cleanup of previous scenario?)")
+				}
+			}
+		}
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
```

Comment:
Is it necessary to cleanup test policies here?

Reply:
No, that's automatic.